x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1365,23 +1393,50 @@ private void unlink(final LFUCacheEntry entry) {
 
         /**
          * Cycles through the {@link LFUCacheEntry} from 0 to max frequency and
-         * tries to evict a chunk if no one is holding onto its resources anymore
+         * tries to evict a chunk if no one is holding onto its resources anymore.
          *
-         * @return the frequency of the evicted entry as integer or -1 if no entry was evicted from cache
+         * Also regularly polls for free regions and thus might steal one in case any become available.
+         *
+         * @return a now free IO region or null if none available.
          */
-        private int maybeEvict() {
+        private SharedBytes.IO maybeEvictAndTake(Runnable evictedNotification) {
             assert Thread.holdsLock(SharedBlobCacheService.this);
             for (int currentFreq = 0; currentFreq < maxFreq; currentFreq++) {
+                // recheck this per freq in case we raced an eviction with an incref'er.
+                SharedBytes.IO freeRegion = freeRegions.poll();
+                if (freeRegion != null) {
+                    return freeRegion;
+                }
                 for (LFUCacheEntry entry = freqs[currentFreq]; entry != null; entry = entry.next) {
-                    boolean evicted = entry.chunk.tryEvict();
-                    if (evicted && entry.chunk.io != null) {
-                        unlink(entry);
-                        keyMapping.remove(entry.chunk.regionKey, entry);
-                        return currentFreq;
+                    boolean evicted = entry.chunk.tryEvictNoDecRef();
+                    if (evicted) {
+                        try {
+                            SharedBytes.IO ioRef = entry.chunk.io;
+                            if (ioRef != null) {
+                                try {
+                                    if (entry.chunk.refCount() == 1) {
+                                        // we own that one refcount (since we CAS'ed evicted to 1)
+                                        // grab io, rely on incref'ers also checking evicted field.
+                                        entry.chunk.io = null;
+                                        assert regionOwners.remove(ioRef) == entry.chunk;
+                                        return ioRef;
+                                    }
+                                } finally {
+                                    unlink(entry);
```

Comment:
This will also stop iterating over other entries with same freq. I think this is OK since we only try to evict 1 CacheFileRegion and take its IO but maybe we can add a `return` statement at the end of the `if (evicted) {` block to make it explicit that no more cache entries will be evaluated?